src/aurora/gff3writer.cpp

```diff
@@ -0,0 +1,610 @@
+	stream.writeUint32LE(structCount);
+	stream.writeUint32LE(fieldOffset);
+	stream.writeUint32LE(fieldCount);
+	stream.writeUint32LE(labelOffset);
+	stream.writeUint32LE(labelCount);
+	stream.writeUint32LE(fieldDataOffset);
+	stream.writeUint32LE(fieldDataCount);
+	stream.writeUint32LE(fieldIndicesOffset);
+	stream.writeUint32LE(fieldIndicesCount);
+	stream.writeUint32LE(listIndicesOffset);
+	stream.writeUint32LE(listIndicesCount);
+
+	// Write structs data.
+	size_t structFieldIndicesIndex = 0;
+	for (size_t i = 0; i < _structs.size(); ++i) {
+		GFF3WriterStructPtr strct = _structs[i];
+
+		// Struct Id.
+		stream.writeUint32LE(strct->getId());
+
+		// Field Index.
+		if (strct->getFieldCount() > 1) {
+			stream.writeUint32LE(structFieldIndicesIndex * 4);
+			structFieldIndicesIndex += strct->getFieldCount();
+		} else {
+			if (strct->getFieldCount() != 0)
+				stream.writeUint32LE(strct->_fieldIndices[0]);
+			else
+				stream.writeUint32LE(0);
+		}
+
+		// Field Count.
+		stream.writeUint32LE(strct->getFieldCount());
+	}
+
+	// Write fields.
+	unsigned int fieldDataIndex = 0;
+	unsigned int listDataIndex = 0;
+	for (size_t i = 0; i < _fields.size(); ++i) {
+		const Field &field = _fields[i];
+		stream.writeUint32LE(field.type);
+		stream.writeUint32LE(field.labelIndex);
+
+		/*
+		 * Determine if this field has simple values (less equal 32 bit) which are written in the field
+		 * or complex values, bigger than 32bit like strings written in the field data section.
+		 */
+		bool simple =
+			field.type == GFF3Struct::kFieldTypeByte ||
+			field.type == GFF3Struct::kFieldTypeChar ||
+			field.type == GFF3Struct::kFieldTypeUint16 ||
+			field.type == GFF3Struct::kFieldTypeUint32 ||
+			field.type == GFF3Struct::kFieldTypeStruct ||
+			field.type == GFF3Struct::kFieldTypeSint16 ||
+			field.type == GFF3Struct::kFieldTypeSint32 ||
+			field.type == GFF3Struct::kFieldTypeFloat ||
+			field.type == GFF3Struct::kFieldTypeList;
+
+		if (simple) {
+			// If the values are simple (less equal 4 bytes) write them to the field.
+			switch (field.type) {
+				case GFF3Struct::kFieldTypeByte:
+				case GFF3Struct::kFieldTypeUint16:
+				case GFF3Struct::kFieldTypeUint32:
+				case GFF3Struct::kFieldTypeStruct:
+					stream.writeUint32LE(field.uint32Value);
+					break;
+				case GFF3Struct::kFieldTypeList:
+					stream.writeUint32LE(listDataIndex * 4);
+					listDataIndex += 1 + _lists[field.uint32Value]->getSize();
+					break;
+				case GFF3Struct::kFieldTypeChar:
+				case GFF3Struct::kFieldTypeSint16:
+				case GFF3Struct::kFieldTypeSint32:
+					stream.writeSint32LE(field.int32Value);
+					break;
+				case GFF3Struct::kFieldTypeFloat:
+					stream.writeIEEEFloatLE(field.floatValue);
+					break;
+				default:
+					throw Common::Exception("Invalid Field type");
+			}
+		} else {
+			// If the values are complex (greater then 4 bytes) write the index to the field data.
+			stream.writeUint32LE(fieldDataIndex);
+			switch (field.type) {
```

Comment:
Isn't that a repeat of lines 61ff?

Move that into a method `getFieldSize()` or something like that.